File: mobile/android/app/src/main/java/club/togetha/app/core/analytics/EngagementTracker.kt
package club.togetha.app.core.analytics

import android.content.Context
import android.util.Log
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.ui.platform.LocalContext
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import club.togetha.app.core.api.Api
import club.togetha.app.core.model.AnalyticsEvent
import club.togetha.app.dataStore
import java.util.UUID
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch

/**
 * Lightweight engagement tracking: a stable per-install session id (DataStore)
 * plus fire-and-forget events batched through ApiClient.trackEvents.
 */
object EngagementTracker {

    private const val TAG = "Engagement"
    private val SESSION_KEY = stringPreferencesKey("session_id")
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    @Volatile private var sessionId: String = "pending"

    fun init(context: Context) {
        val appContext = context.applicationContext
        scope.launch {
            val existing = appContext.dataStore.data.first()[SESSION_KEY]
            sessionId = existing ?: UUID.randomUUID().toString().also { fresh ->
                appContext.dataStore.edit { it[SESSION_KEY] = fresh }
            }
            Log.d(TAG, "session_id=$sessionId")
        }
    }

    fun track(name: String, props: Map<String, String> = emptyMap()) {
        val event = AnalyticsEvent(sessionId, name, props, System.currentTimeMillis())
        scope.launch {
            runCatching { Api.client.trackEvents(listOf(event)) }
                .onFailure { Log.d(TAG, "drop $name: ${it.message}") }
        }
    }

    fun listingView(batchId: String) = track("listing_view", mapOf("batch_id" to batchId))
    fun listingClick(batchId: String) = track("listing_click", mapOf("batch_id" to batchId))
    fun quizStart() = track("quiz_start")
    fun applyStart(batchId: String) = track("apply_start", mapOf("batch_id" to batchId))
}

/** Records a screen_view with dwell seconds when the composable leaves composition. */
@Composable
fun TrackScreen(name: String) {
    val context = LocalContext.current
    DisposableEffect(name) {
        EngagementTracker.init(context)
        val entered = System.currentTimeMillis()
        onDispose {
            val dwellSec = ((System.currentTimeMillis() - entered) / 1000).toString()
            EngagementTracker.track("screen_view", mapOf("screen" to name, "dwell_seconds" to dwellSec))
        }
    }
}

File: mobile/ios/Togetha/Core/Analytics/EngagementTracker.swift
import Foundation
import SwiftUI

/// Lightweight engagement analytics. One session ID per install-session;
/// events flush through `APIClientProtocol.trackEvents` (console in mock,
/// POST /track-event live).
@MainActor
final class EngagementTracker {
    static let shared = EngagementTracker()

    let sessionId = UUID().uuidString
    private var api: any APIClientProtocol = MockAPIClient()
    private var screenStarts: [String: Date] = [:]

    func configure(api: any APIClientProtocol) {
        self.api = api
    }

    // MARK: - Screen dwell

    func screenAppeared(_ name: String) {
        screenStarts[name] = .now
    }

    func screenDisappeared(_ name: String) {
        guard let start = screenStarts.removeValue(forKey: name) else { return }
        let dwell = Date.now.timeIntervalSince(start)
        send(EngagementEvent(
            name: "screen_view", sessionId: sessionId, timestamp: .now,
            screen: name, dwellSeconds: (dwell * 10).rounded() / 10
        ))
    }

    // MARK: - Discrete events

    func listingView(batchId: String) {
        send(EngagementEvent(name: "listing_view", sessionId: sessionId, timestamp: .now, batchId: batchId))
    }

    func listingClick(batchId: String) {
        send(EngagementEvent(name: "listing_click", sessionId: sessionId, timestamp: .now, batchId: batchId))
    }

    func applyStart(batchId: String) {
        send(EngagementEvent(name: "apply_start", sessionId: sessionId, timestamp: .now, batchId: batchId))
    }

    func quizStart(batchId: String?) {
        send(EngagementEvent(name: "quiz_start", sessionId: sessionId, timestamp: .now, batchId: batchId))
    }

    private func send(_ event: EngagementEvent) {
        let api = self.api
        Task { try? await api.trackEvents([event]) }
    }
}

// MARK: - View helper

private struct TrackScreenModifier: ViewModifier {
    let name: String

    func body(content: Content) -> some View {
        content
            .onAppear { EngagementTracker.shared.screenAppeared(name) }
            .onDisappear { EngagementTracker.shared.screenDisappeared(name) }
    }
}

extension View {
    /// Records a `screen_view` event with dwell seconds for this screen.
    func trackScreen(_ name: String) -> some View {
        modifier(TrackScreenModifier(name: name))
    }
}
